// stryker-mocha-runner/src/MochaTestFramework.js
"use strict";
var MochaTestFramework = (function () {
    function MochaTestFramework(settings) {
        this.settings = settings;
    }
    MochaTestFramework.prototype.beforeEach = function (codeFragment) {
        return "beforeEach(function() {\n      " + codeFragment + "\n    });";
    };
    MochaTestFramework.prototype.afterEach = function (codeFragment) {
        return "afterEach(function() {\n      " + codeFragment + "\n    });";
    };
    MochaTestFramework.prototype.filter = function (testIds) {
        return "\n      var mocha = window.mocha || require('mocha');\n      if (window.____mochaAddTest) {\n        mocha.Suite.prototype.addTest = window.____mochaAddTest;\n      } else {\n        window.____mochaAddTest = mocha.Suite.prototype.addTest\n      }\n      var current = 0;\n      var realAddTest = mocha.Suite.prototype.addTest;\n      mocha.Suite.prototype.addTest = function () {\n        if (" + JSON.stringify(testIds) + ".indexOf(current) > -1) {\n          realAddTest.apply(this, arguments);\n        }\n        current++;\n      };\n    ";
    };
    return MochaTestFramework;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = MochaTestFramework;
