// stryker-mocha-runner/src/MochaTestRunner.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var log4js = require('log4js');
var events_1 = require('events');
var test_runner_1 = require('stryker-api/test_runner');
var Mocha = require('mocha');
var StrykerMochaReporter_1 = require('./StrykerMochaReporter');
var log = log4js.getLogger('MochaTestRunner');
var MochaTestRunner = (function (_super) {
    __extends(MochaTestRunner, _super);
    function MochaTestRunner(runnerOptions) {
        _super.call(this);
        this.files = runnerOptions.files;
    }
    MochaTestRunner.prototype.purgeFiles = function () {
        this.files.forEach(function (f) { return delete require.cache[f.path]; });
    };
    MochaTestRunner.prototype.run = function () {
        var _this = this;
        return new Promise(function (resolve, fail) {
            try {
                _this.purgeFiles();
                var mocha_1 = new Mocha({ reporter: StrykerMochaReporter_1.default, bail: true });
                _this.files.filter(function (file) { return file.included; }).forEach(function (f) { return mocha_1.addFile(f.path); });
                try {
                    var runner_1 = mocha_1.run(function (failures) {
                        var result = runner_1.runResult;
                        resolve(result);
                    });
                }
                catch (error) {
                    resolve({
                        status: test_runner_1.RunStatus.Error,
                        tests: [],
                        errorMessages: [error]
                    });
                }
            }
            catch (error) {
                log.error(error);
                fail(error);
            }
        });
    };
    return MochaTestRunner;
}(events_1.EventEmitter));
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = MochaTestRunner;
