// stryker/src/reporters/EventRecorderReporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var log4js = require("log4js");
var path = require("path");
var fileUtils = require("../utils/fileUtils");
var log = log4js.getLogger('EventRecorderReporter');
var DEFAULT_BASE_FOLDER = 'reports/mutation/events';
var EventRecorderReporter = (function () {
    function EventRecorderReporter(options) {
        this.options = options;
        this.allWork = [];
        this.index = 0;
        this.createBaseFolderTask = fileUtils.cleanFolder(this.baseFolder);
    }
    Object.defineProperty(EventRecorderReporter.prototype, "baseFolder", {
        get: function () {
            if (!this._baseFolder) {
                if (this.options['eventReporter'] && this.options['eventReporter']['baseDir']) {
                    this._baseFolder = this.options['eventReporter']['baseDir'];
                    log.debug("Using configured output folder " + this._baseFolder);
                }
                else {
                    log.debug("No base folder configuration found (using configuration: eventReporter: { baseDir: 'output/folder' }), using default " + DEFAULT_BASE_FOLDER);
                    this._baseFolder = DEFAULT_BASE_FOLDER;
                }
            }
            return this._baseFolder;
        },
        enumerable: true,
        configurable: true
    });
    EventRecorderReporter.prototype.writeToFile = function (index, methodName, data) {
        var filename = path.join(this.baseFolder, this.format(index) + "-" + methodName + ".json");
        log.debug("Writing event " + methodName + " to file " + filename);
        return fileUtils.writeFile(filename, JSON.stringify(data));
    };
    EventRecorderReporter.prototype.format = function (input) {
        var str = input.toString();
        for (var i = 10000; i > 1; i = i / 10) {
            if (i > input) {
                str = '0' + str;
            }
        }
        return str;
    };
    EventRecorderReporter.prototype.onSourceFileRead = function (file) {
        var _this = this;
        this.allWork.push(this.createBaseFolderTask
            .then(function () { return _this.writeToFile(_this.index++, 'onSourceFileRead', file); }));
    };
    EventRecorderReporter.prototype.onAllSourceFilesRead = function (files) {
        var _this = this;
        this.allWork.push(this.createBaseFolderTask.then(function () { return _this.writeToFile(_this.index++, 'onAllSourceFilesRead', files); }));
    };
    EventRecorderReporter.prototype.onAllMutantsMatchedWithTests = function (results) {
        var _this = this;
        this.allWork.push(this.createBaseFolderTask.then(function () { return _this.writeToFile(_this.index++, 'onAllMutantsMatchedWithTests', results); }));
    };
    EventRecorderReporter.prototype.onMutantTested = function (result) {
        var _this = this;
        this.allWork.push(this.createBaseFolderTask.then(function () { return _this.writeToFile(_this.index++, 'onMutantTested', result); }));
    };
    EventRecorderReporter.prototype.onAllMutantsTested = function (results) {
        var _this = this;
        this.allWork.push(this.createBaseFolderTask.then(function () { return _this.writeToFile(_this.index++, 'onAllMutantsTested', results); }));
    };
    EventRecorderReporter.prototype.wrapUp = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.createBaseFolderTask];
                    case 1:
                        _a.sent();
                        return [2 /*return*/, Promise.all(this.allWork)];
                }
            });
        });
    };
    return EventRecorderReporter;
}());
exports.default = EventRecorderReporter;
