// stryker-html-reporter/src/HandlebarsModel.js
"use strict";
/**
 * Represents a handlebars result context
 */
var HandlebarsModel = (function () {
    function HandlebarsModel(title, urlPrefix, relativeUrl, totalKilled, totalTimedOut, totalSurvived, totalNoCoverage, totalErrors) {
        this.title = title;
        this.urlPrefix = urlPrefix;
        this.relativeUrl = relativeUrl;
        this.totalKilled = totalKilled;
        this.totalTimedOut = totalTimedOut;
        this.totalSurvived = totalSurvived;
        this.totalNoCoverage = totalNoCoverage;
        this.totalErrors = totalErrors;
        this.totalUndetected = totalSurvived + totalNoCoverage;
        this.totalDetected = totalKilled + totalTimedOut;
        this.totalMutations = this.totalDetected + this.totalUndetected;
        this.totalCoveredMutations = this.totalDetected + totalSurvived;
        this.percentageBasedOnAllCode = Math.floor(this.totalDetected / this.totalMutations * 100) || 0;
        this.percentageBasedOnCoveredCode = Math.floor(this.totalDetected / this.totalCoveredMutations * 100) || 0;
    }
    ;
    return HandlebarsModel;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = HandlebarsModel;
