// stryker-html-reporter/src/SourceFileTreeLeaf.js
"use strict";
var report_1 = require('stryker-api/report');
var HandlebarsModel_1 = require('./HandlebarsModel');
var path = require('path');
var fs = require('fs');
var util = require('./util');
var SourceFileTreeLeaf = (function () {
    function SourceFileTreeLeaf(file, results) {
        if (results === void 0) { results = []; }
        this.file = file;
        this.results = results;
        this.name = path.basename(file.path);
        this.filename = this.name + ".html";
    }
    SourceFileTreeLeaf.prototype.writeFileReport = function (directory) {
        fs.writeFileSync(path.join(directory, this.filename), util.sourceFileTemplate(this), 'utf8');
    };
    SourceFileTreeLeaf.prototype.calculateModel = function (urlPrefix) {
        var killed = 0, timedOut = 0, survived = 0, noCoverage = 0, errors = 0;
        this.results.forEach(function (mutation) {
            switch (mutation.status) {
                case report_1.MutantStatus.Killed:
                    killed++;
                    break;
                case report_1.MutantStatus.TimedOut:
                    timedOut++;
                    break;
                case report_1.MutantStatus.Survived:
                    survived++;
                    break;
                case report_1.MutantStatus.NoCoverage:
                    noCoverage++;
                    break;
                case report_1.MutantStatus.Error:
                    errors++;
                    break;
            }
        });
        this.model = new HandlebarsModel_1.default(this.name, urlPrefix, this.filename, killed, timedOut, survived, noCoverage, errors);
    };
    ;
    return SourceFileTreeLeaf;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = SourceFileTreeLeaf;
