// stryker/src/utils/objectUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var _ = require("lodash");
// Don't use JSON.parse, as it does not allow for regexes or functions, etc
exports.serialize = require('serialize-javascript');
function freezeRecursively(target) {
    Object.freeze(target);
    Object.keys(target).forEach(function (key) {
        if (_.isObject(target[key])) {
            freezeRecursively(target[key]);
        }
    });
}
exports.freezeRecursively = freezeRecursively;
function isPromise(input) {
    return input && typeof input['then'] === 'function';
}
exports.isPromise = isPromise;
function deserialize(serializedJavascript) {
    // Don't use JSON.parse, as it does not allow for regexes or functions, etc
    // tslint:disable
    return eval("(" + serializedJavascript + ")");
    // tslint:enable
}
exports.deserialize = deserialize;
function isErrnoException(error) {
    return typeof error.code === 'string';
}
exports.isErrnoException = isErrnoException;
function errorToString(error) {
    if (!error) {
        return '';
    }
    else if (isErrnoException(error)) {
        return error.name + ": " + error.code + " (" + error.syscall + ") " + error.stack;
    }
    else if (error instanceof Error) {
        var message = error.name + ": " + error.message;
        if (error.stack) {
            return message + "\n" + error.stack.toString();
        }
        else {
            return message;
        }
    }
    else {
        return error.toString();
    }
}
exports.errorToString = errorToString;
function copy(obj, deep) {
    if (deep) {
        return _.cloneDeep(obj);
    }
    else {
        return _.clone(obj);
    }
}
exports.copy = copy;
function wrapInClosure(codeFragment) {
    return "\n    (function (window) {\n      " + codeFragment + "\n    })((Function('return this'))());";
}
exports.wrapInClosure = wrapInClosure;
